substrate/client/transaction-pool/src/fork_aware_txpool/txmempool.rs

```diff
@@ -0,0 +1,260 @@
+	/// is transaction watched
+	watched: bool,
+	//todo: Arc?
+	/// transaction actual body
+	tx: Block::Extrinsic,
+	/// transaction source
+	pub(crate) source: TransactionSource,
+	/// when transaction was revalidated, used to periodically revalidate mem pool buffer.
+	validated_at: AtomicU64,
+}
+
+impl<Block: BlockT> TxInMemPool<Block> {
+	fn is_watched(&self) -> bool {
+		self.watched
+	}
+
+	fn new_unwatched(source: TransactionSource, tx: Block::Extrinsic) -> Self {
+		Self { watched: false, tx, source, validated_at: AtomicU64::new(0) }
+	}
+
+	fn new_watched(source: TransactionSource, tx: Block::Extrinsic) -> Self {
+		Self { watched: true, tx, source, validated_at: AtomicU64::new(0) }
+	}
+
+	pub fn clone_data(&self) -> Block::Extrinsic {
+		self.tx.clone()
+	}
+}
+
+/// Intermediary transaction buffer.
+///
+/// Keeps all the transaction which are potentially valid. Transactions that were finalized or
+/// transaction that are invalid at finalized blocks are removed.
+pub(super) struct TxMemPool<ChainApi, Block>
+where
+	Block: BlockT,
+	ChainApi: graph::ChainApi<Block = Block> + 'static,
+{
+	api: Arc<ChainApi>,
+	//could be removed after removing watched field (and adding listener into tx)
+	listener: Arc<MultiViewListener<ChainApi>>,
+	xts2: RwLock<HashMap<graph::ExtrinsicHash<ChainApi>, Arc<TxInMemPool<Block>>>>,
+	metrics: PrometheusMetrics,
+}
+
+// Clumsy implementation - some improvements shall be done in the following code, use of Arc,
+// redundant clones, naming..., etc...
+impl<ChainApi, Block> TxMemPool<ChainApi, Block>
+where
+	Block: BlockT,
+	ChainApi: graph::ChainApi<Block = Block> + 'static,
+	<Block as BlockT>::Hash: Unpin,
+{
+	pub(super) fn new(
+		api: Arc<ChainApi>,
+		listener: Arc<MultiViewListener<ChainApi>>,
+		metrics: PrometheusMetrics,
+	) -> Self {
+		Self { api, listener, xts2: Default::default(), metrics }
+	}
+
+	pub(super) fn len(&self) -> (usize, usize) {
+		let xts2 = self.xts2.read();
+		let watched_count = xts2.values().filter(|x| x.is_watched()).count();
+		(xts2.len() - watched_count, watched_count)
+	}
+
+	pub(super) fn push_unwatched(&self, source: TransactionSource, xt: Block::Extrinsic) {
+		let hash = self.api.hash_and_length(&xt).0;
+		let unwatched = Arc::from(TxInMemPool::new_unwatched(source, xt));
+		self.xts2.write().entry(hash).or_insert(unwatched);
+	}
+
+	pub(super) fn extend_unwatched(&self, source: TransactionSource, xts: Vec<Block::Extrinsic>) {
+		let mut xts2 = self.xts2.write();
+		xts.into_iter().for_each(|xt| {
+			let hash = self.api.hash_and_length(&xt).0;
+			let unwatched = Arc::from(TxInMemPool::new_unwatched(source, xt));
+			xts2.entry(hash).or_insert(unwatched);
+		});
+	}
+
+	pub(super) fn push_watched(&self, source: TransactionSource, xt: Block::Extrinsic) {
+		let hash = self.api.hash_and_length(&xt).0;
+		let watched = Arc::from(TxInMemPool::new_watched(source, xt));
+		self.xts2.write().entry(hash).or_insert(watched);
```

Comment:
606685df28

Reply:
I am giving it a second thought: actually not overwriting was my intention. Once tx is submitted as watched or unwatched it shall not be changed. Anyway this gonna be developed further because of limits implementation.